quic/s2n-quic-tests/src/tests/ch_callback_connection_info.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+use s2n_quic_core::crypto::tls::ConnectionInfo;
+use s2n_tls::{
+    callbacks::{ClientHelloCallback, ConnectionFuture},
+    error::Error as S2nError,
+};
+use std::{
+    pin::Pin,
+    sync::{Arc, Mutex},
+};
+
+struct TestClientHelloHandle {
+    // The ClientHelloCallback trait requires `&self` as a immutable reference.
+    // We use Arc<Mutex<>> to enable interior mutability - allowing us to mutate the recorded
+    // ConnectionInfo through an immutable reference.
+    recorded_info: Arc<Mutex<Option<ConnectionInfo>>>,
+}
+
+impl TestClientHelloHandle {
+    pub fn new(recorded_info: Arc<Mutex<Option<ConnectionInfo>>>) -> Self {
+        Self { recorded_info }
+    }
+}
+
+impl ClientHelloCallback for TestClientHelloHandle {
+    fn on_client_hello(
+        &self,
+        connection: &mut s2n_tls::connection::Connection,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, S2nError> {
+        let connection_info = connection.application_context::<ConnectionInfo>();
+
+        if let Some(info) = connection_info {
+            *self.recorded_info.lock().unwrap() = Some(*info);
+        }
+
+        Ok(None)
+    }
+}
+
+/// Tests that ConnectionInfo is accessible in the client hello callback and contains
+/// the correct local (server) and remote (client) socket addresses.
+///
+/// This test:
+/// 1. Creates a server with a client hello callback that records ConnectionInfo
+/// 2. Records the actual server and client socket addresses during connection setup
+/// 3. Verifies that the ConnectionInfo captured in the callback matches the expected addresses
+///
+/// Note: Uses interior mutability (Arc<Mutex<>>) to store data from the callback since
+/// ClientHelloCallback requires an immutable reference (&self).
+#[test]
+#[cfg_attr(miri, ignore)]
+fn ch_callback_connection_info_test() {
+    let model = Model::default();
+
+    let ch_callback_handle_inner = Arc::new(Mutex::new(None));
+    let ch_callback_handle_inner_clone = ch_callback_handle_inner.clone();
+
+    let server_local_address = Arc::new(Mutex::new(None));
+    let server_local_address_clone = server_local_address.clone();
+    let server_remote_address = Arc::new(Mutex::new(None));
+    let server_remote_address_clone = server_remote_address.clone();
```

Comment:
I don't think this is necessary you should be able to do:

```suggestion
    let mut server_local_address = None;
    let mut server_remote_address = None;
```